perceval/backends/gerrit.py

```diff
@@ -0,0 +1,297 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2015 Bitergia
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+#
+# Authors:
+#   Alvaro del Castillo San Felix <acs@bitergia.com>
+#
+
+'''Gerrit backend for Perseval'''
+
+
+from datetime import datetime
+import json
+import logging
+import os.path
+import re
+import subprocess
+from time import time
+
+from ..backend import Backend, BackendCommand
+from ..cache import Cache
+from ..errors import BackendError, CacheError
+from ..utils import DEFAULT_DATETIME, str_to_datetime
+
+
+class Gerrit(Backend):
+
+    name = "gerrit"
+
+    def __init__(self, user=None, url=None, nreviews=None,
+                 cache=None, **nouse):
+        super().__init__(cache=cache)
+        self.repository = url
+        self.nreviews = nreviews
+        self.last_review = None  # last review processed
+        self.number_results = None  # last number of results
+        self.client = GerritClient(self.repository, user, nreviews)
+
+    def fetch_from_cache(self):
+        """Fetch the bugs from the cache.
+
+        It returns the bugs stored in the cache object provided during
+        the initialization of the object. If this method is called but
+        no cache object was provided, the method will raise a `CacheError`
+        exception.
+
+        :returns: a generator of items
+
+        :raises CacheError: raised when an error occurs accesing the
+            cache
+        """
+        if not self.cache:
+            raise CacheError(cause="cache instance was not provided")
+
+        cache_items = self.cache.retrieve()
+
+        while True:
+            try:
+                raw_items = next(cache_items)
+                if raw_items is None:
+                    break
+                reviews = self._parse_reviews(raw_items)
+            except StopIteration:
+                break
+
+            for item in reviews:
+                yield item
+
+    def fetch(self, from_date=DEFAULT_DATETIME):
+        self._purge_cache_queue()
+
+        self.last_item = None
+        reviews = self._get_reviews()
+
+        while reviews:
+            review = reviews.pop(0)
+            updated = datetime.fromtimestamp(review['lastUpdated'])
+            if updated <= from_date:
```

Comment:
The reviews in gerrit are retrieved using lastUpdated desc as order. When we find a review that the lastUpdated is older than the from_date, then we don't need to check any more reviews because the rest are older than the from_date (the initial date from which we want reviews).